### PR TITLE
Parameter confidence is optional and CWE's are not always numerical in GitLab SAST

### PR DESCRIPTION
**Pull request by StefanFl**
GitLab SAST doesn't always set the parameter ``confidence`` for a vulnerability, leading to an exception when importing the vulnerabilities.

### CODE REVIEW

**Comment by damiencarol** [comment]
I just reviewed your modification and I have 2 comments here:
1) as the file is loaded with `json.load()` you have dictionaries for the items.
It means that you should use something like that to manage optional attributes:

```Python
scanner_confidence = vuln.get('confidence'], 'Unknown')
```

It seems that the default value is `Unknown`
2) the current code doesn't make any sense.

```Python
scanner_confidence = Finding.get_number_severity(vuln['confidence'])
```
[see here](https://github.com/DefectDojo/django-DefectDojo/pull/3567/files#diff-b50bc2a09a1d04194c4525846f3635f10820bca1ef2ad5acb0ddbd3e430e2248L93)
the function `get_number_severity` has no link to the confidence. **For me it is a bug.**

**Comment by damiencarol** [comment]
Also, @StefanFl could you add another report and keep the old one? Could be good to have the report that makes you exceptions.
If you are ok to provide your report, keep in mind that it's an open-source project so remove the personal details in the report.
Thanks in advance.

File: dojo/tools/gitlab_sast/parser.py [hunk 1]
```diff
@@ -90,7 +90,9 @@ def get_item(vuln, test):
         title = '[{} severity] {}'.format(severity, title)
         severity = 'Info'
     numerical_severity = Finding.get_numerical_severity(severity)
-    scanner_confidence = Finding.get_number_severity(vuln['confidence'])
+    scanner_confidence = None
+    if 'confidence' in vuln:
+        scanner_confidence = Finding.get_number_severity(vuln['confidence'])
```
Comment (damiencarol) [inline dojo/tools/gitlab_sast/parser.py:95]:
```suggestion
        scanner_confidence = vuln.get('confidence', 'Unknown')
```

**Review by damiencarol — CHANGES_REQUESTED** [review]
See my previous comments. Some modifications could be different.

**Comment by damiencarol** [comment]
@StefanFl does this issue is linked to #3534?

**Comment by StefanFl** [comment]
@damiencarol GitLab has 2 different scanners. GitLab SAST is static code analysis (the parser of this PR), while the PR you have referenced parses the results of the GitLab Dependency Scanner. File format for these 2 scanners seems to be pretty similar.

Thank you for the review, will change the files according to your suggestions.

**Comment by StefanFl** [comment]
Interesting one, it was definitely a bug. GitLab SAST can produce a confidence, but the values (search for confidence in https://docs.gitlab.com/ce/development/integrations/secure.html) need to be mapped to the values understood by DefectDojo (see ``Finding.get_scanner_confidence_text()``).

The new report is an amended excerpt from findings for the OWASP Juice Shop, I think this should be allowed.

**Comment by StefanFl** [comment]
Please wait with the review, I think there is another problem with these reports

**Comment by damiencarol** [comment]
Take your time and happy new year!
For the record:
[https://docs.gitlab.com/ce/development/integrations/secure.html#severity-and-confidence](https://docs.gitlab.com/ce/development/integrations/secure.html#severity-and-confidence)

> The confidence ranges from Low to Confirmed, but it can also be Unknown, Experimental or even Ignore if the vulnerability is to be ignored. Valid values are: Ignore, Unknown, Experimental, Low, Medium, High, or Confirmed
>
> Unknown values means that data is unavailable to determine it’s actual value. Therefore, it may be high, medium, or low, and needs to be investigated. We have provided a chart of the available SAST Analyzers and what data is currently available.

**Comment by StefanFl** [comment]
All values are mapped apart from `

-  ``Unkown`` where the confidence is set to ``None`` 
- ``Ignore`` where the confidence is set to ``None`` as well, because there is no respective value in DefectDojo

For the others see the method ``get_confidence_numeric(argument)`` in the parser

**Comment by StefanFl** [comment]
A further test showed, that GitLab sometimes writes non-numerical CWE's which can not be stored in the database. There is a fix for it as well know including a new test case. 

Please re-review the whole PR, thank you.

**Review by damiencarol — APPROVED** [review]
LGTM, high quality modification, this will make Gitlab SAST parser more robust!

One minor feedback, you could have used the feature of subtests to check the items in the unit tests: [https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests)

**Comment by damiencarol** [comment]
@valentijnscholten @madchap it's ready for review.

**Review by valentijnscholten — APPROVED** [review]

**Comment by damiencarol** [comment]
@StefanFl why did you closed this PR ?

**Comment by StefanFl** [comment]
Oops, that wasn't intended, thanks @damiencarol for notifying me.

**Review by valentijnscholten — APPROVED** [review]

**Review by damiencarol — APPROVED** [review]

**Review by madchap — APPROVED** [review]
good job @StefanFl and thanks for the tight review @damiencarol :)

**Pull request merged**